share/wake/lib/system/job.wake

```diff
@@ -168,245 +193,336 @@ export def getPlanHash (plan: Plan): Integer =
-  def build Unit =
-    def visStrings =
-      map getPathName vis
-    def job = create label dir stdin env.implode cmd.implode hash visStrings.implode (if keep then 1 else 0) echo stdout stderr
-    def prefix = "{str pid}.{str (getJobId job)}"
-    def usage = getJobRecord job | getOrElse uusage
-    def output = run job (Pass (RunnerInput label cmd vis env dir stdin res prefix usage))
-    def final _ = match output
-      Fail e =
-        badfinish job e
-      Pass (RunnerOutput inputs outputs (Usage status runtime cputime mem in out)) =
-        def input  = finputs  inputs  | map simplify | implode
-        def output = foutputs outputs | map addhash  | implode
-        finish job input output (implode outputs) status runtime cputime mem in out
-    # Make sure we don't hash files before the job has stopped running
-    def _ = waitJobMerged final job
-    job
-  def confirm abort last job =
-    # notOk cannot be deadcode eliminated thanks to printlnLevel having effects
-    def notOk (Pair name hash) =
-      if hashcode name ==* hash then Unit else
-      if abort # The panic will not be deadcode dropped, because it's an alternative return of the effect-ful notOk
-      # This use of unreachable is not ok!
-      then unreachable "The hashcode of output file '{name}' has changed from {hash} (when wake last ran) to {hashcode name} (when inspected this time). Presumably it was hand edited. Please move this file out of the way. Aborting the build to prevent loss of your data."
-      else printlnLevel logError "Wake was run with '-c' and the hashcode of output file '{name}' has changed, despite being produced from identical inputs. In the prior run, it was {hash} and now it is {hashcode name}. Hashes of dependent jobs using this file will not be checked."
-    def _ = waitJobMerged (\_ map notOk last) job
-    job
-  match keep
-    False = build Unit
-    True  = match (cache dir stdin env.implode cmd.implode hash (map getPathName vis).implode)
-      Pair (job, _) last = confirm True  last job
-      Pair Nil      last = confirm False last (build Unit)
+    def create label dir stdin env cmd signature visible keep echo stdout stderr = prim "job_create"
+    def finish job inputs outputs all_outputs status runtime cputime membytes ibytes obytes =
+        prim "job_finish"
+    def badfinish job error = prim "job_fail_finish"
+    def cache dir stdin env cmd signature visible = prim "job_cache"
+    def signature cmd res fni fno keep = prim "hash"
+    def hash = signature cmd res finputs foutputs keep
+    def build Unit =
+        def visStrings = map getPathName vis
+        def job =
+            create
+            label
+            dir
+            stdin
+            env.implode
+            cmd.implode
+            hash
+            visStrings.implode
+            (
+                if keep then
+                    1
+                else
+                    0
+            )
```

Comment:
This is one of the times the `if` should not be broken up.

Reply:
I agree that this is a case where the `if then else` should all be on one line. But we should probably also have a built-in function for this.

```wake
def booleanToInteger (b: Boolean) : Integer =
    if b then 1 else 0
```

or

```wake
def booleanToInteger (b: Boolean) : Integer = match b
    True = 1
    False = 0
```

The latter function is my preference, because it's easier to read in my opinion with the direct mapping between the boolean and integer values.

Reply:
Looking at the surrounding code, if this `if then else` can't be kept on a single line, I think I'd recommend in code review to not have a `if then else` here and rather either a variable name or simple function application (like `booleanToInteger keep`). I would possibly recommend this regardless. However, this one of those cases where I really wish

```
(
    <stuff>
)
```

could be avoided.